atsign-foundation/at_client_sdk
# fix: Sync missing entries when running with multiple clients and add …

**- What I did**
* **Issue-1:** When multiple clients pump data to the server and attempt to sync the changes, conflicts arise resulting in missing and inconsistencies between clients. Due to conflicts, Client-1 may be missing the commit-Id synced from Client-2 and vice-versa. This occurs because the commit-Id in client-1 is greater than the commit-Id synced from client-2.
  * Please refer to the following [google docs](https://docs.google.com/spreadsheets/d/1JLwe0OupRJSkPC52r9Bo3kshcWF5FRHvLdgHuqj1Jyg/edit?usp=share_link) for more details on the issue.

* **Issue-2:** The ConflictInfo does not set the remote and local value. Following is the git issue: https://github.com/atsign-foundation/at_client_sdk/issues/1016

**- How I did it**
* **For issue 1:** On the client side, currently, we are updating the "lastReceivedServerCommitId" when pushing and pulling the server commit entries. To fix the issue, update the "lastReceivedServerCommitId" only when the client pulls the changes. Because of this fix, "lastReceivedServerCommitId" is not updated on pushing the changes and hence we get synced entries back to the client. If the same exists in the uncommitted entries list the data will be overwritten. To address this issue, on pulling the changes, if an entry with the same key exists in uncommitted entries, skip committing that particular entry to local.
* With new changes, lastReceivedServerCommitId will not be updated when pulling the changes. Hence removed unit tests that assert the same.

* **For Issue 2:** Referring to the "Fix for issue-1", when pulling the changes from the server to the local, if the serverCommitEntry exists in the list of uncommitted entries, then the particular serverCommitEntry is skipped (the entry will not sync to the local secondary), and the lastReceivedServerCommitId is updated. WIth these changes, the control does not go into the section that populates ConflictInfo, because if the serverCommitEntry exists in the uncommitedEntries list, we skip processing the entry. So, refactor the code to populate the conflictInfo and skip syncing the serverCommitEntry to local.

**- How to verify it**
* Updated the unit test to assert the conflict info changes.
* Extended the existing unit test to verify conflictInfo is being set for a shared key 
* Add a new unit test to verify ConflictInfo set errorOrExceptionMessage when exception occurs in populating the ConflictInfo.

**- Description for the changelog**
* Fix: Missing sync entries when multiple clients pump data concurrently.
* Fix: ConflictInfo does not set the remote and local value.
